### ISSUE TITLE
Re-export statements are omitted from output using @babel/plugin-transform-typescript@7.5.0

### ISSUE THREAD

**Issue by elliottsj** [babel/babel#10162]
## Bug Report

**Current Behavior**
Using `@babel/plugin-transform-typescript@7.5.0` via `@babel/preset-typescript@7.3.3`, re-export statements are omitted from the transpiled output.

Repro:

```sh-session
$ npm install @babel/cli @babel/core @babel/preset-typescript
$ echo "export { bar } from './bar';" > script.ts
$ ./node_modules/.bin/babel --presets @babel/preset-typescript script.ts


$ # observe: the above output is empty
```

Note that separate import + export statements transpile properly:

```sh-session
$ echo "import { bar } from './bar'; export { bar };" > script.ts
$ ./node_modules/.bin/babel --presets @babel/preset-typescript script.ts
import { bar } from './bar';
export { bar };

$ # observe: the output is as expected
```

**Input Code**

```js
export { bar } from './bar';
```

**Expected behavior/code**
The re-export statement should be included in the output.

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  "presets": ["@babel/preset-typescript"]
}
```

**Environment**
- Babel version(s):

    ```
    $ npm ls --depth=1                                                                                                                                                                                                                                          (522ms)
    /Users/spencerelliott/Dev/temp
    ├─┬ @babel/cli@7.5.0
    │ ├── chokidar@2.1.6
    │ ├── commander@2.20.0
    │ ├── convert-source-map@1.6.0
    │ ├── fs-readdir-recursive@1.1.0
    │ ├── glob@7.1.4
    │ ├── lodash@4.17.11
    │ ├── mkdirp@0.5.1
    │ ├── output-file-sync@2.0.1
    │ ├── slash@2.0.0
    │ └── source-map@0.5.7
    ├─┬ @babel/core@7.5.0
    │ ├── @babel/code-frame@7.0.0
    │ ├── @babel/generator@7.5.0
    │ ├── @babel/helpers@7.5.0
    │ ├── @babel/parser@7.5.0
    │ ├── @babel/template@7.4.4
    │ ├── @babel/traverse@7.5.0
    │ ├── @babel/types@7.5.0
    │ ├── convert-source-map@1.6.0 deduped
    │ ├── debug@4.1.1 extraneous
    │ ├── json5@2.1.0
    │ ├── lodash@4.17.11 deduped
    │ ├── resolve@1.11.1
    │ ├── semver@5.7.0
    │ └── source-map@0.5.7 deduped
    └─┬ @babel/preset-typescript@7.3.3
      ├── @babel/helper-plugin-utils@7.0.0
      └── @babel/plugin-transform-typescript@7.5.0
    ```

- Node/npm version: Node v10.15.0, npm 6.9.0
- OS: macOS 10.14.5
- Monorepo: no
- How you are using Babel: see above

**Comment by hzoo** [comment]
Wanted to confirm, is this only happening with the TS plugin?

**Comment by elliottsj** [comment]
Yes, I've only noticed this issue with the TS plugin. Pinned to 7.4.5 for now. Thanks!

**Comment by eyousefifar** [comment]
I'm having the same problem

**Comment by Turbo87** [comment]
I just stepped through this after running into the same issue and I can confirm that https://github.com/babel/babel/commit/8d492b159b342cfd1f399e4292e3cb12a895e1d3 is the issue. The only change to that part of the code since v7.4.5 was that particular commit.

**Comment by Wolvereness** [comment]
Just to confirm, 2080042808b18b005fb54aeaf9ab708f64cb1005 and a6392bd636d07c41579ea5ab443152a09943e096 do not have the issue?

**Comment by JLHwung** [comment]
I dug a bit and it seems that a quick fix can be made via guarding against the `exportSpecifier.local` and check `export fromClause` in `ExportNamedDeclaration`

```diff
diff --git a/packages/babel-plugin-transform-typescript/src/index.js b/packages/babel-plugin-transform-typescript/src/index.js
index a768907a5..dcb46231d 100644
--- a/packages/babel-plugin-transform-typescript/src/index.js
+++ b/packages/babel-plugin-transform-typescript/src/index.js
@@ -99,9 +99,12 @@ export default declare(
           // remove export declaration if it's exporting only types
           if (
             path.node.specifiers.length > 0 &&
-            !path.node.specifiers.find(exportSpecifier =>
-              path.scope.hasOwnBinding(exportSpecifier.local.name),
-            )
+            !path.node.specifiers.find(
+              exportSpecifier =>
+                exportSpecifier.local &&
+                path.scope.hasOwnBinding(exportSpecifier.local.name),
+            ) &&
+            !path.node.source
           ) {
             path.remove();
           }

```

**Comment by pixelass** [comment]
I noticed that all my default exports were missing.

```
export default = "foo"
```


with `7.5.0`

```
var _default = "foo";
;
```

with `7.4.5`

```
const _default = "foo";
var _default2 = _default;
exports.default = _default2;
;
```

**Comment by JLHwung** [comment]
@Wolvereness Hi I just figure out a quick fix of this issue and it works well on my cases. Are you also working on a fix? If so I would leave it as-is and stop here.

```diff
diff --git a/packages/babel-plugin-transform-typescript/src/index.js b/packages/babel-plugin-transform-typescript/src/index.js
index a768907a5..79ca00102 100644
--- a/packages/babel-plugin-transform-typescript/src/index.js
+++ b/packages/babel-plugin-transform-typescript/src/index.js
@@ -98,9 +98,10 @@ export default declare(
         ExportNamedDeclaration(path) {
           // remove export declaration if it's exporting only types
           if (
+            !path.node.source &&
             path.node.specifiers.length > 0 &&
             !path.node.specifiers.find(exportSpecifier =>
-              path.scope.hasOwnBinding(exportSpecifier.local.name),
+              path.scope.hasBinding(exportSpecifier.local.name),
             )
           ) {
             path.remove();
@@ -109,7 +110,10 @@ export default declare(
 
         ExportSpecifier(path) {
           // remove type exports
-          if (!path.scope.hasOwnBinding(path.node.local.name)) {
+          if (
+            !path.parent.source &&
+            !path.scope.hasBinding(path.node.local.name)
+          ) {
             path.remove();
           }
         },
@@ -118,7 +122,7 @@ export default declare(
           // remove whole declaration if it's exporting a TS type
           if (
             t.isIdentifier(path.node.declaration) &&
-            !path.scope.hasOwnBinding(path.node.declaration.name)
+            !path.scope.hasBinding(path.node.declaration.name)
           ) {
             path.remove();
           }

```

**Comment by Wolvereness** [comment]
106bb41d47fddb92d3c2d7ff14e99bedeffaa57d (master as of now):

* No difference

8d492b159b342cfd1f399e4292e3cb12a895e1d3 (my rewrite commit):

* Input:
  ```typescript
  export default "foo";
  ```

  Output *unchanged from parent commit*:
  ```javascript
  export default "foo";
  ```

* Input:
  ```typescript
  export { bar } from './bar';
  ```

  Output *exact commit of change*:
  ```
  ```

a6392bd636d07c41579ea5ab443152a09943e096 *Exact commit of change*:

* Input:
  ```typescript
  export default "foo";
  ```

  Output:
  ```javascript
  ```

* Input:
  ```typescript
  export { bar } from './bar';
  ```

  Output:
  ```javascript
  ```

2080042808b18b005fb54aeaf9ab708f64cb1005 *No change from parent*:

* Input:
  ```typescript
  export default "foo";
  ```

  Output:
  ```javascript
  export default "foo";
  ```

* Input:
  ```typescript
  export { bar } from './bar';
  ```

  Output:
  ```javascript
  export { bar } from './bar';
  ```

**Comment by shrinktofit** [comment]
+1

**Comment by qiqiboy** [comment]
This issue seems still exists in v7.5.1.

**Comment by JoshRobertson** [comment]
In my case:
In 7.5.0 compilation was successful but would throw an error at runtime.
In 7.5.1 compilation fails with an `export was not found` error, which is the same error I had with 7.4.5. So I'm stuck on 7.4.4 still.

**Comment by nicolo-ribaudo** [comment]
@qiqiboy @JoshRobertson Can you share your code?

**Comment by JoshRobertson** [comment]
@nicolo-ribaudo 

sample repo here: https://github.com/JoshRobertson/babel-typescript-issue

**Comment by nicolo-ribaudo** [comment]
@JoshRobertson Babel can't handle re-exports of TypeScript types, since we don't do cross-file analysis and by looking at this code:
```js
export { Avatar, AvatarProps } from "./Avatar";
```
it's impossible to know that `AvatarProps` is a type.

Since usages of `AvatarProps` will be removed anyway, you can safely ignore the webpack warning: it tells you that something can't be exported, but it isn't being used anyway.

**Comment by qiqiboy** [comment]
`.babelrc`
```json
{
  "plugins": ["react-hot-loader/babel", "@babel/plugin-transform-typescript"]
}
```

Input:
```typescript
export defualt 'Foo';
```
Output:
```javascript
const _default = 'Foo';
; // what???
```
May this is related with `react-hot-loader`? I think it's because the updates from `>=v7.5.0` make some third-party plugins incompatible.

**Comment by nicolo-ribaudo** [comment]
Yes, it is a problem with `react-hot-loader`. A plugin is responsible of maintaining scope information up-to-date, and Babel 7.5.0 brought to the surface that bug in `react-hot-loader/babel` because it uses scope information to remove exports.
I'm surprised that this didn't come up before, since we use scope info in many other places.

Would you be willing to open a PR to that package (or open an issue)?
https://github.com/gaearon/react-hot-loader/blob/master/src/babel.dev.js#L118
`insertBefore` returns an array of ast paths, and that plugin should call `path.scope.registerDeclaration(resultOfInsertBefore[0])`

**Comment by theKashey** [comment]
So that stuff was there for years, and everybody was just lucky :)

**Comment by theKashey** [comment]
If someone could verify https://github.com/gaearon/react-hot-loader/pull/1293 with a patch @nicolo-ribaudo proposed - I'll issue a new version of React-Hot-Loader.
I was unable to find documentation for `registerDeclaration` and have no tests previously red, so you are the ones, who could say "👍"

**Comment by sciyoshi** [comment]
Seems to be fixed for me by updating `react-hot-loader` to 4.12.5 with https://github.com/gaearon/react-hot-loader/pull/1293 Thank you @theKashey @nicolo-ribaudo!

**Comment by nicolo-ribaudo** [comment]
We just released Babel 7.5.2. Please check that in your lockfile it contains v7.5.2 of `@babel/plugin-transform-typescript`!
